usim/py/events.py

```diff
@@ -0,0 +1,642 @@
+        if delay < 0:
+            raise ValueError("'delay' must not be negative")
+        super().__init__(env)
+        self._delay = delay
+        self._fixed_value = value
+        env.schedule(self._trigger_timeout())
+
+    async def _trigger_timeout(self):
+        await (time + self._delay)
+        self.succeed(self._fixed_value)
+
+    def __repr__(self):
+        return f'<usim.py.{self.__class__.__name__} delay={self._delay}>'
+
+
+class Initialize:
+    """
+    Internal initializer for :py:class:`~.Process`
+
+    Since this event is internal and not used by 'usim.py',
+    it is not implemented. Instantiating this class
+    unconditionally raises an error.
+    """
+    def __init__(self, *args, **kwargs):
+        raise NotEmulatedError(self.__doc__)
+
+
+class InterruptQueue:
+    """
+    Internal helper to manage interrupts for :py:class:`~.Process`
+    """
+    __slots__ = '__usimpy_flag__', '_causes', 'ok', 'defused'
+
+    def __init__(self):
+        self.__usimpy_flag__ = Flag()
+        self._causes = []
+        self.ok = False
+        self.defused = True
+
+    def __bool__(self):
+        return bool(self._causes)
+
+    @property
+    def value(self):
+        """The next :py:exc:`Interrupt`; this is a destructive action"""
+        return Interrupt(self.pop())
+
+    def push(self, cause):
+        """Add a new interrupt with ``cause``"""
+        self._causes.append(cause)
+        if not self.__usimpy_flag__:
+            self.__usimpy_flag__._value = True
+            self.__usimpy_flag__.__trigger__()
+
+    def pop(self):
+        result = self._causes.pop(0)
+        if not self._causes:
+            self.__usimpy_flag__._value = False
+        return result
+
+
+class Process(Event[V]):
+    """
+    A concurrently running process which triggers as an event on completion
+
+    Processes wrap around generators, which in turn may ``yield`` events to
+    wait for them. This implements cooperative multitasking, allowing another
+    process to run while others wait. Each :py:class:`~.Process` acts as an
+    event which is automatically triggered when the underlying generator completes.
+
+    .. code:: python3
+
+        def clock(env: Environment, sound):
+            for _ in range(60):
+                print(sound)
+                yield env.timeout(1)
+
+        def clocks(env: Environment):
+            env.process('tick')
+            env.process('tack')
+            yield env.process('TOCK')
+
+    .. hint::
+
+        **Migrating to μSim**
+
+        When migrating from :py:class:`~.Process`, activities are expressed as
+        ``async def`` coroutines instead of generators. Activities do not need
+        an environment - it is implicitly available. An activity can ``await``
+        a :term:`notification`, concurrently iterate using ``async for``
+        and manage resources using ``async with``.
+
+        .. code:: python3
+
+            async def clock(sound):
```

Comment:
Is it on purpose to have two clock functions? The example below uses this clock function and not the other above. So by just copying the code users might run into problems.